rajashared/cpu-shmem-ltimes.cpp

```diff
@@ -0,0 +1,564 @@
+
+        }
+
+      }
+    }
+  }
+
+
+
+  timer.stop();
+  printf("LTimes took %lf seconds using bare loops and RAJA::Views\n",
+      timer.elapsed());
+
+
+
+  // Check correctness
+  if(debug){
+
+    size_t errors = 0;
+    double total_error = 0.;
+    for (IZone z(0); z < num_zones; ++z) {
+      for (IGroup g(0); g < num_groups; ++g) {
+        for (IMoment m(0); m < num_moments; ++m) {
+          double total = 0.0;
+          for (IDirection d(0); d < num_directions; ++d) {
+            double val = ell(m, d) * psi(d, g, z);
+            total += val;
+          }
+          if(std::abs(total-phi(m,g,z)) > 1e-9){
+            ++ errors;
+          }
+          total_error += std::abs(total-phi(m,g,z));
+        }
+      }
+    }
+    if(errors == 0){
+      printf("  -- no errors (%e)\n", total_error);
+    }
+    else{
+      printf("  -- failed : %ld errors\n", (long)errors);
+    }
+  }
+
+}
+
+
+
+
+void runLTimesRajaNested(bool debug,
+                          Index_type num_moments,
+                          Index_type num_directions,
+                          Index_type num_groups,
+                          Index_type num_zones)
+{
+
+  using namespace RAJA::nested;
+
+  // psi[direction, group, zone]
+  using PsiView = RAJA::TypedView<double, Layout<3, Index_type, 2>, IDirection, IGroup, IZone>;
+
+  // phi[moment, group, zone]
+  using PhiView = RAJA::TypedView<double, Layout<3, Index_type, 2>, IMoment, IGroup, IZone>;
+
+  // ell[moment, direction]
+  using EllView = RAJA::TypedView<double, Layout<2, Index_type, 1>, IMoment, IDirection>;
+
+
+  // allocate data
+  // phi is initialized to all zeros, the others are randomized
+  std::vector<double> ell_data(num_moments * num_directions);
+  std::vector<double> psi_data(num_directions * num_groups * num_zones);
+  std::vector<double> phi_data(num_moments * num_groups * num_zones);
+
+
+  // randomize data
+  for (size_t i = 0; i < ell_data.size(); ++i) {
+    ell_data[i] = i; //drand48();
+  }
+
+  for (size_t i = 0; i < psi_data.size(); ++i) {
+    psi_data[i] = 2*i; //drand48();
+  }
+
+  for (size_t i = 0; i < phi_data.size(); ++i) {
+    phi_data[i] = 0; //drand48();
+  }
+
+
+  double *d_ell = &ell_data[0];
+  double *d_phi = &phi_data[0];
+  double *d_psi = &psi_data[0];
+
+
+
+  // create views on data
+  std::array<RAJA::idx_t, 2> ell_perm {{0, 1}};
+  EllView ell(
+      d_ell,
+      make_permuted_layout({{num_moments, num_directions}}, ell_perm));
+
+  std::array<RAJA::idx_t, 3> psi_perm {{0, 1, 2}};
+  PsiView psi(
+      d_psi,
+      make_permuted_layout({{num_directions, num_groups, num_zones}}, psi_perm));
+
+  std::array<RAJA::idx_t, 3> phi_perm {{0, 1, 2}};
+  PhiView phi(
+      d_phi,
+      make_permuted_layout({{num_moments, num_groups, num_zones}}, phi_perm));
+
+
+
+  using Pol = RAJA::nested::Policy<
+    For<0, loop_exec,
+      For<1, loop_exec,
+        For<2, loop_exec,
+          For<3, simd_exec, Lambda<0>>
```

Comment:
@ajkunen  , Intel won't be able to vectorize this.

Reply:
why?

Reply:
It looks like you are accumulating a value (reduction like operation), when building I also get the "warning #13379: loop was not vectorized with "simd"